core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -698,7 +700,16 @@ public MergeWriterNode plan(Merge merge)
                     Expression original = mergeCase.getSetExpressions().get(index);
                     Expression setExpression = coerceIfNecessary(analysis, original, original);
                     subPlan = subqueryPlanner.handleSubqueries(subPlan, setExpression, analysis.getSubqueries(merge));
-                    rowBuilder.add(subPlan.rewrite(setExpression));
+                    Expression rewritten = subPlan.rewrite(setExpression);
+                    if (nonNullableColumnHandles.contains(dataColumnHandle)) {
+                        int fieldIndex = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(dataColumnHandle), "Could not find fieldIndex for non nullable column");
+                        ColumnSchema columnSchema = dataColumnSchemas.get(fieldIndex);
+                        String columnName = columnSchema.getName();
+                        Expression coalesced = new CoalesceExpression(rewritten, new Cast(failFunction(metadata, session, INVALID_ARGUMENTS, "Assigning NULL to non-null MERGE target table column " + columnName), toSqlType(columnSchema.getType())));
```

Comment:
coalesced -> checked

or, I'd perhaps simply reassign `rewritten` and keep `rowBuilder.add(rewritten);` after the `if`.